STM32: Replace init structures for drivers that use the timers

gautierg-st opened the pull request:
This PR applies changes to several STM32 drivers that use timers: counter, PWM and QDEC.
It removes the usage of the various initialization structures functions that are located in the source files of the LL. They are replaced by several calls to simple static inline functions from the LL header file.
This avoids the compilation of the LL source file during the build process, and clarifies what exactly is initialized.

It has been tested on our test bench for counter and PWM, but not for QDEC (except for compilation), since we have no hardware to test it. I would greatly appreciate if someone could test the encoder part and share the results.

gautierg-st committed:
drivers: counter: stm32: remove init struct

Remove the calls to LL_TIM_StructInit and LL_TIM_Init in the counter
driver. This avoids calling functions from stm32xxxx_ll_tim.c.
They are replaced by a set of simpler functions from the header file.

Also replaces some macros that used constants coming from the HAL. Use
an equivalent coming purely from the LL.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: pwm stm32: remove init struct

Remove the calls to LL_TIM_xx_StructInit and LL_TIM_xx_Init in the PWM
driver. This avoids calling functions from stm32xxxx_ll_tim.c.
They are replaced by a set of simpler functions from the header file.

OC Init in particular is much simpler now. The init structure needed to be
filled out with the complementary channel (if it existed), even though its
configuration didn't change.
The new init is much more direct and only touches what needs to be
modified.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: pwm: stm32: reduce pointer level from cfg->timer to timer

This change reduces the level of pointer indirection, which minimizes
repeated dereferencing and helps reduce the overall code size, in the same
way as commit 48e326a5520ec884f38f6a7ac4e88c59a01ceb95 for UART.

This also fixes the type complimentary -> complementary.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: sensor: st: qdec: remove init struct

Remove the calls to LL_TIM_ENCODER_StructInit and LL_TIM_ENCODER_Init
in the QDEC driver. This avoids calling functions from stm32xxxx_ll_tim.c.
They are replaced by a set of simpler functions from the header file.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

mathieuchopstm commented on drivers/counter/counter_ll_stm32_timer.c:
```diff
@@ -413,17 +412,28 @@ static int counter_stm32_init_timer(const struct device *dev)
-	init.CounterMode = LL_TIM_COUNTERMODE_UP;
-	init.Autoreload = counter_get_max_top_value(dev);
-	init.ClockDivision = LL_TIM_CLOCKDIVISION_DIV1;
+	if (IS_TIM_COUNTER_MODE_SELECT_INSTANCE(timer)) {
+		LL_TIM_SetCounterMode(timer, LL_TIM_COUNTERMODE_UP);
+	}
+
+	LL_TIM_SetAutoReload(timer, counter_get_max_top_value(dev));
+
+	if (IS_TIM_CLOCK_DIVISION_INSTANCE(timer)) {
+		LL_TIM_SetClockDivision(timer, LL_TIM_CLOCKDIVISION_DIV1);
+	}
```
Nit: I'd group all conditional blocks together
```suggestion
	LL_TIM_SetAutoReload(timer, counter_get_max_top_value(dev));
	
	if (IS_TIM_COUNTER_MODE_SELECT_INSTANCE(timer)) {
		LL_TIM_SetCounterMode(timer, LL_TIM_COUNTERMODE_UP);
	}

	if (IS_TIM_CLOCK_DIVISION_INSTANCE(timer)) {
		LL_TIM_SetClockDivision(timer, LL_TIM_CLOCKDIVISION_DIV1);
	}
```

mathieuchopstm commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -288,106 +289,58 @@ static int pwm_stm32_set_cycles(const struct device *dev, uint32_t channel,
 	bool is_inverted = (flags & PWM_POLARITY_MASK) == PWM_POLARITY_INVERTED;
-	LL_TIM_IC_InitTypeDef ic;
+	uint32_t ll_channel = ch2ll[channel - 1];
+	uint32_t ll_complementary_channel = ch2ll[complementary_channel[channel] - 1];
```
There is a confusing and an anti-pattern at play here:
* If I understand correctly, `complementary_channel[0]` is never used (to allow not needing to do `complementary_channel[ch - 1]`): this should be documented
* As far as I can see, values from the `complementary_channel` array are always `- 1`'d: in this case, it would be wiser to have pre-decremented values in the array and remove the unneeded subtractions

Both of these ("[0] is unused" and "values are pre-decremented") should be documented in the comment above the array's definition.


As a more general remark, the `[0] is unused` optimization could be applied to all the other arrays of the driver - I'm pretty sure the footprint cost would be recouped, and it ought to be faster (and more readable).

gautierg-st committed:
drivers: pwm: stm32: remork complementary channel logic

Align the definition of the complementary channels to the normal channels.
That way, it is consistent for all arrays, we use channel-1 as index, and
no other operation is necessary.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

mathieuchopstm commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -45,10 +45,10 @@ enum capture_state {
  * that is used to capture the end of the pulse.
  */
-static const uint32_t complimentary_channel[] = {0, 2, 1, 4, 3};
+static const uint32_t complementary_channel[] = {2, 1, 4, 3};
```
Shouldn't this account for pre-decrement now?
```suggestion
static const uint32_t complementary_channel[] = {1, 0, 3, 2};
```

gautierg-st replied:
Done
